batch/batch/cloud/azure/driver/pricing.py

```diff
@@ -0,0 +1,165 @@
+                 effective_start_date: int,
+                 effective_end_date: Optional[int] = None,
+                 ):
+        self.machine_type = machine_type
+        self.preemptible = preemptible
+        self.region = region
+        self.cost_per_hour = cost_per_hour
+        self.effective_start_date = effective_start_date
+        self.effective_end_date = effective_end_date
+
+
+class AzureDiskPrice(AzurePrice):
+    def __init__(self,
+                 disk_name: str,
+                 redundancy_type: str,
+                 size_gib: int,
+                 region: str,
+                 cost_per_month: float,
+                 effective_start_date: int,
+                 effective_end_date: Optional[int] = None,
+                 ):
+        self.disk_name = disk_name
+        self.redundancy_type = redundancy_type
+        self.size_gib = size_gib
+        self.region = region
+        self.cost_per_month = cost_per_month
+        self.effective_start_date = effective_start_date
+        self.effective_end_date = effective_end_date
+
+    @property
+    def disk_product(self):
+        return f'{self.disk_product}_{self.redundancy_type}'
+
+    @property
+    def disk_type(self):
+        return self.disk_product[0]
+
+    @property
+    def cost_per_gib_month(self):
+        return self.cost_per_month / self.size_gib
+
+
+async def get_vm_prices_by_region(pricing_client: aioazure.AzurePricingClient,
+                                  region: str,
+                                  machine_types: List[str],
+                                  ) -> List[AzureVMPrice]:
+    prices: List[AzureVMPrice] = []
+    seen_vm_names: Dict[str, str] = {}
+
+    filter_args = ['(serviceName eq \'Virtual Machines\')',
+                   f'(armRegionName eq \'{region}\')']
+
+    machine_type_filter_args = []
+    for machine_type in machine_types:
+        machine_type_filter_args.append(f'(armSkuName eq \'{machine_type}\')')
+
+    machine_type_filter = '(' + ' or '.join(machine_type_filter_args) + ')'
+    filter_args.append(machine_type_filter)
+
+    filter = ' and '.join(filter_args)
+    async for data in pricing_client.list_prices(filter=filter):
+        if 'Windows' in data['productName'] or data['type'] != 'Consumption' or 'Low Priority' in data['skuName']:
+            continue
+        assert data['unitOfMeasure'] == '1 Hour' and data['currencyCode'] == 'USD', data
+
+        sku_name = data['skuName']
+        machine_type = data['armSkuName']
+        preemptible = ('Spot' in sku_name)
+        vm_cost_per_hour = float(data['retailPrice'])
+
+        start_date = parse_timestamp_msecs(data['effectiveStartDate'])
+        end_date = parse_timestamp_msecs(data.get('effectiveEndDate'))
+
+        if sku_name in seen_vm_names:
+            seen_data = seen_vm_names[sku_name]
+            log.exception(f'already seen pricing for vm {sku_name}; {seen_data} vs {data}; aborting')
+            raise Exception()
```

Comment:
If we raise, why do we log? The caller is obligated to either handle the exception or log the issue. I want to reduce the amount of repeated exception logs we get. Also, shouldn't the exception include the error information?